paddle/cinn/ir/group_schedule/tactic/arrange_storage_tactic.cc

```diff
@@ -0,0 +1,407 @@
+            << block_store_range.Min() << " : " << block_store_range.Max()
+            << "]";
+    VLOG(6) << "block_load_range of [" << iter_values_of_load[i] << "] = ["
+            << block_load_range.Min() << " : " << block_load_range.Max() << "]";
+    std::optional<bool> is_block_sub_set =
+        block_load_range.ProveSubSet(block_store_range);
+    if (!is_block_sub_set.has_value() || !is_block_sub_set.value()) {
+      VLOG(6) << "load range of a cuda block is not sub set of store";
+      return CudaAxisType::kCudaBlock;
+    }
+
+    IntSet thread_store_range = Evaluate(iter_values_of_store[i],
+                                         cuda_block_thread_fixed_var_of_store,
+                                         store_var_domain);
+    IntSet thread_load_range = Evaluate(iter_values_of_load[i],
+                                        cuda_block_thread_fixed_var_of_load,
+                                        load_var_domain);
+    VLOG(6) << "thread_store_range of [" << iter_values_of_store[i] << "] = ["
+            << thread_store_range.Min() << " : " << thread_store_range.Max()
+            << "]";
+    VLOG(6) << "thread_load_range of [" << iter_values_of_load[i] << "] = ["
+            << thread_load_range.Min() << " : " << thread_load_range.Max()
+            << "]";
+    std::optional<bool> is_thread_sub_set =
+        thread_load_range.ProveSubSet(thread_store_range);
+    if (!is_thread_sub_set.has_value() || !is_thread_sub_set.value()) {
+      VLOG(6) << "load range of a cuda thread is not sub set of store";
+      return CudaAxisType::kCudaThread;
+    }
+  }
+
+  return std::nullopt;
+}
+
+ArrangeStorageTactic::ArrangeStorageTactic(
+    const std::unordered_set<std::string>& output_names)
+    : output_names_(output_names) {}
+
+void ArrangeStorageTactic::Apply(ir::IRSchedule* sch,
+                                 const std::string& block_id) {
+  ir::Expr store_block = sch->GetBlock(block_id);
+  ir::Expr root_block = sch->GetRootBlock(store_block);
+  ir::Expr store = *ir::ir_utils::CollectIRNodesWithoutTensor(
+                        store_block,
+                        [&](const ir::Expr* x) { return x->As<ir::Store>(); },
+                        true)
+                        .begin();
+
+  VarToForMap var2for_map =
+      analyzer::CollectVarToForMap({root_block}, sch->GetAllBlocks());
+
+  // Traverse load nodes to check if there are loads that cross cuda blocks or
+  // threads
+  std::vector<std::pair<Expr, Expr>> loads_and_blocks =
+      analyzer::GetConsumerLoadsAndSBlocks(store_block, root_block);
+
+  ir::MemoryType memory_type = ir::MemoryType::GPULocal;
+  for (const auto& load_and_block : loads_and_blocks) {
+    ir::Expr load = load_and_block.first;
+    ir::Expr load_block = load_and_block.second;
+    std::optional<CudaAxisType> cross_type =
+        AnalyzeCrossType(var2for_map, store, load, store_block, load_block);
+    if (!cross_type.has_value()) {
```

Comment:
This seems to me that the logic is:

If we don't know cross Block or cross Thread, set to local

Should we do a safer logic? Such as

If we know it won't cross Block or cross Thread, set to local
if we know cross Block, throw error
else set to shared

Note: the difference of two logic is the handling of unknown case.

Reply:
The cross_type without a value indicates that it does not cross any threads or blocks.
In the AnalyzeCrossType, for safety reasons, if it is uncertain whether it crosses blocks, it will be considered as crossing blocks, and the same applies to cross threads.